### ISSUE TITLE
Generate netCDF-4 format file from cdl files instead of netCDF-classic format files

### ISSUE THREAD

**Issue by neumannd** [ioos/compliance-checker#725]
# Current situation:

When a cdl file is provided to the compliance checker, a netCDF-classic file is created via `ncgen`. 

Code (`suite.py`, l.669ff):

``` python
        if '.cdl' in cdl_path:  # it's possible the filename doesn't have the .cdl extension
            ds_str = cdl_path.replace('.cdl', '.nc')
        else:
            ds_str = cdl_path + '.nc'
        subprocess.call(['ncgen', '-o', ds_str, cdl_path])
        return ds_str
```

# Problem

The cdl file might contain netCDF-4 features that do not exist in the netCDF-classic format -- e.g. additional [primitive data types](https://www.unidata.ucar.edu/software/netcdf/docs/netcdf_utilities_guide.html#cdl_data_types). The usage of one/some of the additional/new data types causes `ncgen` to fail when a netCDF-classic file is create from the cdl file.

# Reproduction of the issue

Please find attached to cdl files:

* [example_with_netcdf4_feature.txt](https://github.com/ioos/compliance-checker/files/4155871/example_with_netcdf4_feature.txt)
* [example_without_netcdf4_feature.txt](https://github.com/ioos/compliance-checker/files/4155872/example_without_netcdf4_feature.txt)

``` bash
# succeeds (create netCDF classic file)
ncgen example_without_netcdf4_feature.txt -o example_without_netcdf4_feature.nc
ncgen -k nc3 example_without_netcdf4_feature.txt -o example_without_netcdf4_feature.nc

# fails (create netCDF classic file from cdl file with netCDF4 features)
ncgen example_with_netcdf4_feature.txt -o example_with_netcdf4_feature.nc
ncgen -k nc3 example_with_netcdf4_feature.txt -o example_with_netcdf4_feature.nc

# solution that succeeds for file `example_with_netcdf4_feature.txt`
ncgen -k nc4 example_with_netcdf4_feature.txt -o example_with_netcdf4_feature.nc
```

# Solution

We generate a netCDF4 file instead of a netCDF-classic file by modifying the code from

``` python
subprocess.call(['ncgen', '-o', ds_str, cdl_path])
```

to

``` python
subprocess.call(['ncgen', '-k', 'nc4', '-o', ds_str, cdl_path])
```

If this modification is of interest, I can create a pull request for this.

**Comment by neumannd** [comment]
Alternatively, one could try to create a netCDF classic file, first, and, in case of an error, try to create a netCDF-4 file as follows:

```
        # generate netCDF classic file
        iostat = subprocess.call(['ncgen', '-k', 'nc4', '-o', ds_str, cdl_path])
        if iostat != 0:
          # if not successfull, create netCDF-4 file
          print('netCDF Classic file could not be generated from cdl file ' +
                cdl_path + '. Trying to create netCDF-4 file.')
          iostat = subprocess.call(['ncgen', '-k', 'nc4', '-o', ds_str, cdl_path])
          if iostat != 0:
            # If neither a netCDF Classic nor a netCDF-4 file could be
            # created, throw an error.
            raise RuntimeError('`ncgen` could not generate netCDF file ' +
                               '(classic and netCDF4) from provided cdl' +
                               ' file: ' + cdl_path + '. Generation of ' +
                               'the netCDF4 terminated with the error code' +
                               str(iostat) + '.')
```

I could also provide three new test cases to check the proper behavior.

**Comment by daltonkell** [comment]
I am fully in support of adding this enhancement! It may be advantageous to try and create a NetCDF-4 file by default, as this is the newer, enhanced format, and then fall back on the NetCDF Classic format. Perhaps @benjwadams has an opinion on this.

Looking forward to the pull request.

**Comment by neumannd** [comment]
Accidentally, both ncgen-calls created a netCDF-4 file. Here is a modified version (first netCDF-4; fall-back netCDF classic):

``` python
# generate netCDF-4 file
iostat = subprocess.call(['ncgen', '-k', 'nc4', '-o', ds_str, cdl_path])
if iostat != 0:
    # if not successfull, create netCDF classic file
    print('netCDF-4 file could not be generated from cdl file ' +
        cdl_path + '. Trying to create netCDF Classic file.')
    iostat = subprocess.call(['ncgen', '-k', 'nc3', '-o', ds_str, cdl_path])
    if iostat != 0:
    # If neither a netCDF Classic nor a netCDF-4 file could be
    # created, throw an error.
    raise RuntimeError('`ncgen` could not generate netCDF file ' +
                        '(classic and netCDF4) from provided cdl' +
                        ' file: ' + cdl_path + '. Generation of ' +
                        'the netCDF file terminated with the error' +
                        ' code ' + str(iostat) + '.')
```

**Comment by benjwadams** [comment]
Implemented by #729, closing.
